drivers/flash/flash_ipuc/flash_ipuc.c

```diff
@@ -0,0 +1,458 @@
+	DEVICE_DEFINE(flash_nrf_ipuc_##x, "nrf_ipuc_" #x, NULL, /* init */                         \
+		      NULL,					/* pm */                           \
+		      &ipuc_context_data_##x, NULL,		/* config */                       \
+		      POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT, &nrf_ipuc_api)
+
+#define DEFINE_NRF_IPUC_REF(x, _) &__device_flash_nrf_ipuc_##x
+
+static int nrf_ipuc_read(const struct device *dev, off_t offset, void *data, size_t len)
+{
+	struct ipuc_context *ctx = NULL;
+
+	if ((dev == NULL) || (data == NULL)) {
+		return -EINVAL;
+	}
+
+	ctx = (struct ipuc_context *)dev->data;
+	if (!ctx->read_access) {
+		return -EACCES;
+	}
+
+	LOG_DBG("read: %p:%zu", (void *)offset, len);
+
+	if (offset + len > ctx->size) {
+		return -ENOMEM;
+	}
+
+	memcpy(data, (void *)(ctx->address + offset), len);
+
+	return 0;
+}
+
+static int nrf_ipuc_write(const struct device *dev, off_t offset, const void *data, size_t len)
+{
+	struct ipuc_context *ctx = NULL;
+	bool last_chunk = false;
+
+	if (dev == NULL) {
+		return -EINVAL;
+	}
+
+	ctx = (struct ipuc_context *)dev->data;
+
+	LOG_DBG("write: %p:%zu", (void *)offset, len);
+
+	if (len == 0) {
+		last_chunk = true;
+	}
+
+	if (offset + len > ctx->size) {
+		return -ENOMEM;
+	}
+
+	suit_plat_err_t plat_ret =
+		suit_ipuc_write(&ctx->component_id, offset, (uintptr_t)data, len, last_chunk);
+
+	if (plat_ret != SUIT_PLAT_SUCCESS) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int nrf_ipuc_erase(const struct device *dev, off_t offset, size_t size)
+{
+	static uint8_t erase_block[WRITE_BLOCK_SIZE] = {ERASE_VALUE};
+	suit_plat_err_t plat_ret = SUIT_PLAT_SUCCESS;
+	struct ipuc_context *ctx = NULL;
+
+	if (dev == NULL) {
+		return -EINVAL;
+	}
+
+	ctx = (struct ipuc_context *)dev->data;
+
+	if (offset + size > ctx->size) {
+		return -ENOMEM;
+	}
+
+	LOG_DBG("erase: %p:%zu", (void *)offset, size);
+	while (size > 0) {
+		plat_ret = suit_ipuc_write(&ctx->component_id, offset, (uintptr_t)erase_block,
+					   sizeof(erase_block), false);
+		offset += sizeof(erase_block);
+		size -= sizeof(erase_block);
+	}
+
+	if (plat_ret != SUIT_PLAT_SUCCESS) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static const struct flash_parameters *nrf_ipuc_get_parameters(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	static const struct flash_parameters parameters = {
+		/* Pretend to support single-byte access to avoid read/update/write cycles
+		 * in case of unaligned write requests.
+		 */
+		.write_block_size = 1,
+		.erase_value = ERASE_VALUE,
+		.caps = {.no_explicit_erase = true},
+	};
+
+	return &parameters;
+}
+
+static const struct flash_driver_api nrf_ipuc_api = {
+	.read = nrf_ipuc_read,
+	.write = nrf_ipuc_write,
+	.erase = nrf_ipuc_erase,
+	.get_parameters = nrf_ipuc_get_parameters,
+};
+
+static bool read_access_check(suit_manifest_role_t role)
+{
+	switch (role) {
+#ifdef CONFIG_SOC_NRF54H20_CPUAPP
+	case SUIT_MANIFEST_APP_ROOT:
+	case SUIT_MANIFEST_APP_RECOVERY:
```

Comment:
Is this really enough? What about Secure vs Non-secure context?

Reply:
I think we've never discussed SUIT and S/NS split.
Either way - IPUC read access is needed only for cache over IPUC functionality, that (in most cases) will address NS part of the app.